### PR TITLE
Add support for macOS hosts

### PR DESCRIPTION
**Pull request by dgsga**
This initial changeset adds support for macOS hosts which do not have a built-in afp server (macOS Big Sur onwards) or which do not support AFP 2.2, the version required to connect to vintage Macs running classic Mac OS. In summary:

is compatible with modern Intel and Apple Silicon based Macs
fixes compilation errors when using the macOS native compiler. Apple's version of clang is picky about header orders, it likes the system headers to be iterated before the atalk and same-dir headers.
fixes macro to enable Zeroconf support on macOS
fixes EA macro to ensure sys EA support
fixes the lock file location on macOS
fixes db3 and ssl macros to ensure OpenSSL and Berkeley DB dependencies installed via Homebrew are found
fixes PAM macro to enable authentication on macOS
disables ACL and dtrace support as Apple implementations are currently not supported
installs a macOS LaunchDaemon to start netatalk on boot and restart from the command line when required
requires dependencies to be installed via Homebrew:
brew install berkeley-db libevent libgcrypt mysql libressl openssl@1.1 pkg-config
What is currently unsupported in this changeset (but possible, proof of concept please see my fork https://github.com/dgsga/netatalk):

building documentation using docbook (WIP)
DBUS support / afpstats (WIP)
I have tested these code changes extensively on my Macs over a two year period. I have also tested the PR branch on Debian 11 to check for inadvertent regressions, none found yet. Please test if you have Mac hardware available

### CODE REVIEW

**Comment by dgsga** [comment]
Re-submitted after source repo cleanup.

File: macros/db3-check.m4 [hunk 1]
```diff
@@ -86,7 +86,7 @@ AC_DEFUN([AC_NETATALK_PATH_BDB],[
 if test "x$bdb_required" = "xyes"; then
     trybdbdir=""
     dobdbsearch=yes
-    bdb_search_dirs="/usr/local /usr"
+    bdb_search_dirs="/usr/local /usr /usr/local/opt/berkeley-db /opt/homebrew/opt/berkeley-db"
```
Comment (rdmark) [inline macros/db3-check.m4:89]: Rather than hard coding these obscure paths, would it be possible to ask macOS users to pass the `--with-bdb=/path/to/bdb/installation/` configure parameter instead?

File: macros/ssl-check.m4 [hunk 2]
```diff
@@ -45,7 +45,7 @@ dnl	compile_ssl=no
 
 	if test "$tryssl" = "yes"; then
 		AC_MSG_CHECKING([for SSL])
-		for ssldir in "" $tryssldir /usr /usr/local/openssl /usr/$atalk_libname/openssl /usr/local/ssl /usr/$atalk_libname/ssl /usr/local /usr/pkg /opt /opt/openssl /usr/local/ssl ; do
+		for ssldir in "" $tryssldir /usr /usr/local/openssl /usr/$atalk_libname/openssl /usr/local/ssl /usr/$atalk_libname/ssl /usr/local /usr/pkg /opt /opt/openssl /usr/local/ssl /usr/local/opt/openssl@1.1 /opt/homebrew/opt/openssl@1.1; do
```
Comment (rdmark) [inline macros/ssl-check.m4:48]: Similarly to above, there is the `--with-ssl-dir=PATH` configure option that should work in lieu of hard coding here. Or is there a reason this won't work?

File: libatalk/bstring/bstrlib.c [hunk 3]
```diff
@@ -2737,7 +2737,7 @@ struct genBstrList g;
 #define START_VSNBUFF (256)
 #else
 
-#ifdef __GNUC__
+#if defined(__GNUC__) && !defined(__clang__)
```
Comment (rdmark) [inline libatalk/bstring/bstrlib.c:2740]: I think this inline prototype is bitrotted. Just now I put up a PR to remove it. https://github.com/Netatalk/netatalk/pull/316

File: macros/netatalk.m4 [hunk 4]
```diff
@@ -486,10 +493,17 @@ AC_ARG_ENABLE(shell-check,
 dnl Check for optional initscript install
 AC_DEFUN([AC_NETATALK_INIT_STYLE], [
     AC_ARG_WITH(init-style,
-                [  --with-init-style       use OS specific init config [[redhat-sysv|redhat-systemd|suse-sysv|suse-systemd|gentoo-openrc|gentoo-systemd|netbsd|debian-sysv|debian-systemd|solaris|openrc|systemd]]],
+                [  --with-init-style       use OS specific init config [[redhat-sysv|redhat-systemd|suse-sysv|suse-systemd|gentoo-openrc|gentoo-systemd|netbsd|debian-sysv|debian-systemd|solaris|openrc|systemd|macos-launchd]]],
```
Comment (rdmark) [inline macros/netatalk.m4:496]: The html documentation should to be updated with this new init-style option: https://github.com/Netatalk/netatalk/blob/main/doc/manual/install.xml

File: macros/netatalk.m4 [hunk 5]
```diff
@@ -52,6 +52,13 @@ AC_DEFUN([AC_NETATALK_DTRACE], [
     [WDTRACE=$withval],
     [WDTRACE=auto]
   )
+  case "$host_os" in
+  *darwin*)
+  if test x"$WDTRACE" = x"auto" ; then
+  WDTRACE=no
+  fi
+  ;;
+  esac
```
Comment (rdmark) [inline macros/netatalk.m4:61]: Might be worth leaving a comment here to explain why dtrace is being disabled.

File: macros/db3-check.m4 (hunk 1 above)
Comment (dgsga) [reply macros/db3-check.m4:89]: It certainly would be possible but these paths are not obscure if you use a Mac as your primary machine for compilation of open source software.

File: macros/ssl-check.m4 (hunk 2 above)
Comment (dgsga) [reply macros/ssl-check.m4:48]: As above...

File: libatalk/bstring/bstrlib.c (hunk 3 above)
Comment (dgsga) [reply libatalk/bstring/bstrlib.c:2740]: Great, thanks for this. This particular patch has been discarded from my PR branch...

File: macros/netatalk.m4 (hunk 4 above)
Comment (dgsga) [reply macros/netatalk.m4:496]: Will add and upload to my PR branch.

File: macros/netatalk.m4 (hunk 5 above)
Comment (dgsga) [reply macros/netatalk.m4:61]: As above!

File: macros/netatalk.m4 (hunk 4 above)
Comment (dgsga) [reply macros/netatalk.m4:497]: **Update:** done

File: macros/netatalk.m4 (hunk 5 above)
Comment (dgsga) [reply macros/netatalk.m4:62]: **Update:** done

File: macros/db3-check.m4 (hunk 1 above)
Comment (rdmark) [reply macros/db3-check.m4:89]: The location is obscure in the sense that it's whatever the homebrew team decided to use at one time, and might be subject to change in the future. Another package manager might become popular and take over, like what happened when macports lost popularity.

I think the --with-bdb override exists just for this kind of usecase -- an unofficial package manager and non-standard library locations.

Of course I do see the convenience factor for macos users, but I wanted to discuss the sustainable way to manage this macro. It's a bit of a slippery slope to try to support every non standard location...

File: macros/ssl-check.m4 (hunk 2 above)
Comment (rdmark) [reply macros/ssl-check.m4:48]: Would it at least be possible to not lock openssl to v1.1 in the macro? The world is rapidly moving towards openssl 3...

File: macros/db3-check.m4 (hunk 1 above)
Comment (rdmark) [reply macros/db3-check.m4:89]: Case in point: OmniOS (an OpenSolaris descendent) installs the Berkeley libs in /opt/ooce so configuring netatalk with `--with-bdb=/opt/ooce` becomes mandatory.

Comment (dgsga) [reply macros/db3-check.m4]: Fair do's, I have removed the SSL and db3 macro mods from the changes. By the way, OpenSSL3 breaks DHX authentication when logging in from classic Mac OS clients because it enforces a minimum DH modulus size of 512 bits. DHX (otherwise known as DHCAST128) uses a modulus size of 128 bits so you have to compile against OpenSSL 1.1 or libreSSL to login successfully.

File: macros/ssl-check.m4 (hunk 2 above)
Comment (dgsga) [reply macros/ssl-check.m4]: Please see above...

File: macros/db3-check.m4 (hunk 1 above)
Comment (dgsga) [reply macros/db3-check.m4]: I have also updated install.xml to prompt users to specify a non standard SSL install location in the same way as the manual already does for BDB

Comment (rdmark) [reply macros/db3-check.m4]: Looking good, thanks for making those changes.

Well noted on the breaking change in openssl3; TIL.

FWIW I have the habit of disabling DHX and using clrtxt for auth with classic Mac OS clients, DHX2 with macOS clients, since DHX was buggy on ARM for the longest time (works now I think.)

AFAICT the upcoming Debian Bookworm has dropped openssl1.1 altogether now so it's a bridge that we have to cross sooner or later...

**Comment by rdmark** [comment]
A few thoughts for after this is merged:
- I see you have a github workflow set up in your fork to build on macOS. I would like to add this to the CI checks, in parallel with the Ubuntu job.
- A wiki page that covers the required homebrew packages, and other quirks of macOS would be helpful.

**Comment by dgsga** [comment]
> A few thoughts for after this is merged:
> 
> * I see you have a github workflow set up in your fork to build on macOS. I would like to add this to the CI checks, in parallel with the Ubuntu job.

Fell free to copy and paste.

**Comment by dgsga** [comment]
> A few thoughts for after this is merged:
> 
> * A wiki page that covers the required homebrew packages, and other quirks of macOS would be helpful.

Done!

File: macros/db3-check.m4 (hunk 1 above)
Comment (dgsga) [reply macros/db3-check.m4]: LibreSSL is a good alternative and is likely to maintain support for older protocols for longer.

Comment (rdmark) [reply macros/db3-check.m4]: Good point... netatalk3 already supports libressl I think.

**Comment by rdmark** [comment]
> > A few thoughts for after this is merged:
> > 
> > * A wiki page that covers the required homebrew packages, and other quirks of macOS would be helpful.
> 
> Done, will be updated as required!

Looks good! Just a few thoughts...
- Link to homebrew's installation page, in case the installation procedure changes in the future
- Don't you need also xcode and/or Apple's development packages installed first?
- How do you launch netatalk after installing?

**Comment by rdmark** [comment]
Got netatalk running on my MBP M1 Max! That was very seamless. Great job smoothing out all of the bumps.

**Comment by rdmark** [comment]
As a side note, I keep running into error -50 on Mac clients when copying files with EA onto shared volumes. This is not specific to macOS hosts, but thought I'd mention it in case you've seen this too and know how to avoid it. I bet it's some kind of permissions / configuration issue.

**Comment by rdmark** [comment]
@slowfranklin This changeset has turned out less complex / intrusive than I anticipated, so I'm leaning towards merging it sooner rather than later and include it in the next patch release. Thoughts?

**Comment by rdmark** [comment]
I enabled and did a squashed merge for this changeset, for a cleaner commit log.

@dgsga Thanks for your hard work making this happen!
